File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/economy/EconomyService.java
package top.likoslupus.cellulosesz.api.economy;

import java.math.BigDecimal;
import java.util.Collection;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public interface EconomyService {

    BigDecimal balance(UUID uuid);

    default String format(BigDecimal amount) {
        return amount.toPlainString();
    }

    CompletableFuture<TransactionResult> deposit(
            UUID uuid,
            BigDecimal amount,
            TransactionCause cause
    );

    CompletableFuture<TransactionResult> withdraw(
            UUID uuid,
            BigDecimal amount,
            TransactionCause cause
    );

    CompletableFuture<TransactionResult> setBalance(
            UUID uuid,
            BigDecimal amount,
            TransactionCause cause
    );

    CompletableFuture<TransactionResult> transfer(
            UUID from,
            UUID to,
            BigDecimal amount,
            TransactionCause cause
    );

    CompletableFuture<TransactionResult> transferMany(
            UUID from,
            Collection<UUID> recipients,
            BigDecimal amountEach,
            TransactionCause cause
    );

    default List<BalanceEntry> topBalances(int limit) {
        return topBalances(limit, BalanceFilter.all());
    }

    List<BalanceEntry> topBalances(int limit, BalanceFilter filter);

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/economy/EconomyService.kt
package top.likoslupus.cellulosesz.api.economy

import java.math.BigDecimal
import java.util.*
import java.util.concurrent.CompletableFuture

public interface EconomyService {

    public fun balance(uuid: UUID): BigDecimal

    public fun format(amount: BigDecimal): String = amount.toPlainString()

    public fun deposit(
        uuid: UUID,
        amount: BigDecimal,
        cause: TransactionCause
    ): CompletableFuture<TransactionResult>

    public fun withdraw(
        uuid: UUID,
        amount: BigDecimal,
        cause: TransactionCause
    ): CompletableFuture<TransactionResult>

    public fun setBalance(
        uuid: UUID,
        amount: BigDecimal,
        cause: TransactionCause
    ): CompletableFuture<TransactionResult>

    public fun transfer(
        from: UUID,
        to: UUID,
        amount: BigDecimal,
        cause: TransactionCause
    ): CompletableFuture<TransactionResult>

    public fun transferMany(
        from: UUID,
        recipients: Collection<UUID>,
        amountEach: BigDecimal,
        cause: TransactionCause
    ): CompletableFuture<TransactionResult>

    public fun topBalances(limit: Int): List<BalanceEntry> =
        topBalances(limit, BalanceFilter.all())

    public fun topBalances(
        limit: Int,
        filter: BalanceFilter
    ): List<BalanceEntry>

}
